### PR TITLE
Check if the lid might block due to missing commit.

### PR DESCRIPTION
**Pull request by baldersheim**
If so pill back and reschedule.
Rescheduling will give a busy loop, but that is rare and find.

@toregge PR
A test is need, and ditto for lidspace job.

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp [hunk 1]
```diff
@@ -125,14 +125,15 @@ BucketMoveJob::moveDocuments(DocumentBucketMover &mover,
         bucketGuard = _frozenBuckets.acquireExclusiveBucket(mover.getBucket());
         if (! bucketGuard) {
             maybeDelayMover(mover, mover.getBucket());
-            return;
+            return true;
         }
     }
     assert(mover.getBucket() == bucketGuard->getBucket());
-    mover.moveDocuments(maxDocsToMove);
+    if ( ! mover.moveDocuments(maxDocsToMove)) return false;
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp:132]: Consider adding braces to fix style here.

File: searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp [hunk 2]
```diff
@@ -334,7 +335,10 @@ BucketMoveJob::run()
     if (isBlocked()) {
         return true; // indicate work is done, since node state is bad
     }
-    scanAndMove(200, 1);
+    /// Returning false here will immediately post the job back on the executor. This will give a busy loop,
+    /// but this is considered fine as it is very rare and it will be intermingled with multiple feed operations.
+    if ( ! scanAndMove(200, 1) ) return false;
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp:340]: Consider adding braces to fix style.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 3]
```diff
@@ -19,14 +19,15 @@ namespace proton {
...
 typedef IDocumentMetaStore::Iterator Iterator;
 
-void
+bool
 DocumentBucketMover::moveDocument(DocumentIdT lid,
                                   const document::GlobalId &gid,
                                   Timestamp timestamp)
 {
-    Document::SP doc(_source->retriever()->getFullDocument(lid).release());
+    if ( _source->lidNeedsCommit(lid) ) return false;
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:27]: Consider adding braces to fix style.

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp [hunk 4]
```diff
@@ -131,8 +134,9 @@ DocumentBucketMover::moveDocuments(size_t maxDocsToMove)
         setBucketDone();
     }
     for (const MoveKey & key : toMove) {
-        moveDocument(key._lid, key._gid, key._timestamp);
+        if ( ! moveDocument(key._lid, key._gid, key._timestamp)) return false;
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp:137]: Consider adding braces to fix style.

File: searchcore/src/vespa/searchcore/proton/server/maintenancedocumentsubdb.cpp [hunk 5]
```diff
@@ -38,4 +41,10 @@ MaintenanceDocumentSubDB::clear()
     _feed_view.reset();
 }
 
+bool
+MaintenanceDocumentSubDB::lidNeedsCommit(search::DocumentIdT lid) const {
+    return (_pendingLidsForCommit &&
```
Comment (toregge) [inline searchcore/src/vespa/searchcore/proton/server/maintenancedocumentsubdb.cpp:46]: Consider adding `!= nullptr` for a more standard style with pointer checks.

**Review by toregge — APPROVED** [review]

File: searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/bucketmovejob.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp (hunk 3 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp (hunk 4 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/documentbucketmover.cpp]: Fixed

File: searchcore/src/vespa/searchcore/proton/server/maintenancedocumentsubdb.cpp (hunk 5 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/server/maintenancedocumentsubdb.cpp]: Fixed

**Pull request merged**